examples/gno.land/p/demo/timelock/timelock.gno

```diff
@@ -0,0 +1,181 @@
+//
+// Example Usage:
+//
+//  import "gno.land/p/demo/timelock"
+//  import "gno.land/p/demo/accesscontrol"
+//
+//  Initialize timelock utility with an AVL tree and access control.
+//  timestamps := avl.NewTree()
+//  adminRole := accesscontrol.NewRoleData("admin", std.Address("admin-address"))
+//  timeLockUtil := timelock.NewTimeLockUtil(timestamps, adminRole, 30)
+//
+//  Schedule an operation with a delay of 60 seconds.
+//  id := seqid.ID()
+//  timeLockUtil.Schedule(id, 60)
+//
+//  Check if an operation is pending.
+//  isPending := timeLockUtil.IsPending(id)
+//
+//  Execute the operation when it is pending.
+//  if timeLockUtil.IsPending(id) {
+//      timeLockUtil.Execute(id)
+//  }
+//
+//  Update the minimum delay for future operations.
+//  timeLockUtil.UpdateDelay(45)
+
+package timelock // import "gno.land/p/demo/timelock"
+
+import (
+	"std"
+	"strconv"
+	"time"
+
+	"gno.land/p/demo/ufmt"
+
+	"gno.land/p/demo/accesscontrol"
+	"gno.land/p/demo/avl"
+	"gno.land/p/demo/seqid"
+)
+
+// Marks operations completed
+const DoneTimestamp = 1
+
+// Represents the status of a planned operation
+type OperationState int
+
+const (
+	Unset OperationState = iota
+	Pending
+	Ready
+	Done
+)
+
+// TimeLockUtil stores the necessary parameters for the timelock operations
+type TimeLockUtil struct {
+	timestamps    *avl.Tree
+	accessControl *accesscontrol.RoleData
+	minDelay      uint64
+}
+
+// New instance of TimeLockUtil
+func MustNewTimeLockUtil(timestamps *avl.Tree, accessControl *accesscontrol.RoleData, minDelay uint64) (*TimeLockUtil, error) {
+	if timestamps == nil || accessControl == nil {
+		return nil, ufmt.Errorf("timestamps and accesscontrol values must be different from nil")
+	}
+	return &TimeLockUtil{
+		timestamps:    timestamps,
+		accessControl: accessControl,
+		minDelay:      minDelay,
+	}, nil
+}
+
+// Schedules an operation to be carried out after a minimum delay
+func (tl *TimeLockUtil) Schedule(id seqid.ID, delay uint64) error {
+	if delay < tl.minDelay {
+		return ufmt.Errorf("timelockutil: Schedule: insufficient delay")
+	}
+	if tl.timestamps.Has(id.Binary()) {
+		return ufmt.Errorf("timelockutil: Schedule: operation already scheduled")
+	}
+	tl.timestamps.Set(id.Binary(), uint64(time.Now().Unix())+delay)
+	std.Emit("CallScheduled", "id", id.String(), "delay", strconv.FormatInt(int64(delay), 10))
+	return nil
+}
+
+// Remove operation
+func (tl *TimeLockUtil) Remove(id seqid.ID) {
+	tl.timestamps.Remove(id.Binary())
+	std.Emit("Removed", "id", id.String())
+}
+
+// Cancels a planned operation
+func (tl *TimeLockUtil) Cancel(id seqid.ID) error {
+	if !tl.IsPending(id) {
+		ufmt.Errorf("timelock: Cancel: operation not pending")
+	}
+	tl.timestamps.Remove(id.Binary())
+	std.Emit("Cancelled", "id", id.String())
+	return nil
+}
+
+// Executes a pending operation
+func (tl *TimeLockUtil) Execute(id seqid.ID) error {
+	if !tl.IsPending(id) {
+		ufmt.Errorf("timelock: Execute: operation not pending")
```

Comment:
For all the spots where `ufmt.Error` is used, it may be better to use `errors.New` when there is no formatting happening.

Reply:
- https://github.com/gnolang/gno/pull/2307/commits/8da6e065c93489c876ae2ca331fb774f5937f758